micropython/mesptool/mesptool.py

```diff
@@ -0,0 +1,233 @@
+# This file is part of the MicroPython project, http://micropython.org/
+#
+# The MIT License (MIT)
+#
+# Copyright (c) 2022 Ibrahim Abdelkader <iabdalkader@openmv.io>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+#
+# A minimal esptool implementation to communicate with ESP32 ROM bootloader.
+# Note this tool does Not support advanced features, other ESP chips or stub loading.
+# This is only meant to be used for updating the U-blox Nina module firmware.
+
+import os
+import struct
+from machine import Pin
+from machine import UART
+from micropython import const
+from time import sleep
+
+
+class mesptool:
+    FLASH_ID = const(0)
+    FLASH_BLOCK_SIZE = const(64 * 1024)
+    FLASH_SECTOR_SIZE = const(4 * 1024)
+    FLASH_PAGE_SIZE = const(256)
+
+    CMD_SYNC = const(0x08)
+    CMD_CHANGE_BAUDRATE = const(0x0F)
+    CMD_SPI_ATTACH = const(0x0D)
+    CMD_SPI_FLASH_MD5 = const(0x13)
+    CMD_SPI_FLASH_PARAMS = const(0x0B)
+    CMD_SPI_FLASH_BEGIN = const(0x02)
+    CMD_SPI_FLASH_DATA = const(0x03)
+    CMD_SPI_FLASH_END = const(0x04)
+
+    ESP_ERRORS = {
+        0x05: "Received message is invalid",
+        0x06: "Failed to act on received message",
+        0x07: "Invalid CRC in message",
+        0x08: "Flash write error",
+        0x09: "Flash read error",
+        0x0A: "Flash read length error",
+        0x0B: "Deflate error",
+    }
+
+    def __init__(
+        self, reset=3, gpio0=2, uart_id=1, uart_tx=Pin(8), uart_rx=Pin(9), log_enabled=False
+    ):
```

Comment:
As per most other drivers, I think it's better to get the user to pass in Pin and UART objects.  Then it's more general.  Eg the user can pass in a USB_VCP object, or a Signal instead of Pin, etc.

You can still do this to configure the pins:
```py
reset_pin.init(reset_pin.OUT)
```

Reply:
On STM32 it's has to work that way -- UARTs cannot be constructed from pins.